base/strings/faces.jl

```diff
@@ -0,0 +1,779 @@
+name, `default` is added to both [`DEFAULT_FACES`](@ref) and (a copy of) to
+[`FACES`](@ref), with the [`FACES`](@ref) value returned.
+
+Should the face `name` already exist, `nothing` is returned.
+"""
+function addface!(name::Symbol, default::Face)
+    if !haskey(DEFAULT_FACES, name)
+        DEFAULT_FACES[name] = default
+        FACES[name] = copy(default)
+    end
+end
+
+"""
+    resetfaces!()
+
+Reset [`FACES`](@ref) to the default value (`DEFAULT_FACES`).
+"""
+function resetfaces!()
+    global FACES = copy(DEFAULT_FACES)
+end
+
+"""
+    resetfaces!(name::Symbol)
+
+Reset the face `name` to its default value, which is returned.
+If the face `name` does not exist, nothing is done and `nothing` returned.
+In the unlikely event that the face `name` does not have a default value,
+it is deleted, a warning message is printed, and `nothing` returned.
+"""
+function resetfaces!(name::Symbol)
+    if !haskey(FACES, name)
+    elseif haskey(DEFAULT_FACES, name)
+        FACES[name] = copy(DEFAULT_FACES[name])
+    else # This shouldn't happen
+        delete!(FACES, name)
+        println(stderr,
+                """! The face $name was reset, but it had no default value, and so has been deleted instead!",
+                     This should not have happened, perhaps the face was added without using `addface!`?""")
+    end
+end
+
+"""
+    withfaces(f, kv::Pair...)
+
+Execute `f` with [`FACES`](@ref) temporarily modified by zero or more `:name =>
+val` arguments `kv`. `withfaces` is generally used via the `withfaces(kv...) do
+... end` syntax. A value of `nothing` can be used to temporarily unset an
+environment variable (if it is set). When `withfaces` returns, the original
+[`FACES`](@ref) has been restored.
+
+    !!! warning
+    Changing faces is not thread-safe.
+"""
+function withfaces(f, keyvals::Pair{Symbol, Union{Face, Symbol, Nothing}}...)
+    old = Dict{Symbol, Union{Face, Nothing}}()
+    for (name, face) in keyvals
+        old[name] = get(FACES, face, nothing)
+        if face isa Face
+            FACES[name] = face
+        elseif face isa Symbol
+            FACES[name] = get(FACES, face, Face())
+        elseif haskey(FACES, name)
+            delete!(FACES, name)
+        end
+    end
+    try f()
+    finally
+        for (name, face) in old
+            if isnothing(face)
+                delete!(FACES, name)
+            else
+                FACES[name] = face
+            end
+        end
+    end
+end
+
+"""
+    withfaces(f, altfaces::Dict{Symbol, Face})
+
+Execute `f` with [`FACES`](@ref) temporarily swapped out with `ALTFACES`
+When `withfaces` returns, the original [`FACES`](@ref) has been restored.
+
+    !!! warning
+    Changing faces is not thread-safe.
+"""
+function withfaces(f, altfaces::Dict{Symbol, Face})
+    global FACES
+    oldfaces, FACES = FACES, altfaces
+    try f()
+    finally
+        FACES = oldfaces
+    end
+end
```

Comment:
Is there a reason you don't have a `rethrow` here?

```suggestion
function withfaces(f, altfaces::Dict{Symbol, Face})
    global FACES
    oldfaces, FACES = FACES, altfaces
    try 
        f()
    catch err
        FACES = oldfaces
        rethrow(err)
    finally
        FACES = oldfaces
    end
end

Reply:
It's not needed/